soccer/planning/InterceptPlanner.cpp

```diff
@@ -0,0 +1,109 @@
+#include "InterceptPlanner.hpp"
+
+#include <Configuration.hpp>
+#include <motion/TrapezoidalMotion.hpp>
+#include <Constants.hpp>
+
+#include "RRTPlanner.hpp"
+#include "MotionInstant.hpp"
+
+
+namespace Planning{
+std::unique_ptr<Path> InterceptPlanner::run(PlanRequest& planRequest) {
+    const InterceptCommand& command = 
+        dynamic_cast<const InterceptCommand&>(*planRequest.motionCommand);
+
+    // Start state for the specified robot
+    const MotionInstant& startInstant = planRequest.start;
+
+    // All the max velocity / acceleration constraints for translation / rotation
+    const MotionConstraints& motionConstraints = planRequest.constraints.mot;
+
+    SystemState& systemState = planRequest.systemState;
+    const Ball& ball = systemState.ball;
+
+    // Time for ball to hit target point
+    // Target point is projected into ball vel line
+    Geometry2d::Point targetPosOnLine;
+    RJ::Seconds ballToPointTime = ball.estimateTimeTo(command.target, &targetPosOnLine) - RJ::now();
+
+    // Try to use the same point if it's pretty close
+    // Improves consistency overall
+    if ((targetPosOnLine - prevPathTarget).mag() < Robot_Radius / 2) {
+        targetPosOnLine = prevPathTarget;
+    } else {
+        prevPathTarget = targetPosOnLine;
+    }
+
+    // vector from robot to target
+    Geometry2d::Point botToTarget = (targetPosOnLine - startInstant.pos);
+    // Normalized vector from robot to target
+    Geometry2d::Point botToTargetNorm = botToTarget.normalized();
+
+    // Max speed we can reach given the distance to target and constant acceleration
+    // If we don't constrain the speed, there is a velocity discontinuity in the middle of the path
+    double maxSpeed = std::min(startInstant.vel.mag() + sqrt(2*motionConstraints.maxAcceleration*botToTarget.mag()),
+                               motionConstraints.maxSpeed);
+
+    // Try to shortcut
+    // If we are almost in the way of the ball already, just stay still
+    // Saves some frames trying to compute that
+    // Also is much more consistent
```

Comment:
This being "more consistent" kind of worries me about the operation in normal conditions. Is this a particular failure case or is the planner sketchy in general?

Reply:
Due to noise in the ball velocity it jitters a little bit. Since the jitter changes the intercept target point and usually the ball is pretty close to the robot, the robot launches full speed "towards" the target point and doesn't stand still like it should.